### DISCUSSION TITLE
how to create a SLR pulse by pypulseq

### DISCUSSION THREAD

**Discussion by qizixinge** [imr-framework/pypulseq#226] [Q&A] [answered]
I want to replace original sinc pulse by SLR pulse, but I don't know how to. I made 2 attempts on write_gre.py, but neither was successful.  

![first_try](https://github.com/user-attachments/assets/91806803-1d30-4181-aaa5-b02120976dbe)

![first_problem](https://github.com/user-attachments/assets/ff82d0be-5438-41bc-8bab-6d958d8f0990)
![second_try](https://github.com/user-attachments/assets/9989beaf-9d13-40ec-bbe3-85c75d45aa0d)

![second_problem](https://github.com/user-attachments/assets/d00c7238-e7c8-4264-8abc-c652edea0e11)

So how can I get SLR pulse with 3mm slice thickness by code? It would be even better if apodization=0.42 could be set

**Comment by btasdelen** [comment]
Hi @qizixinge. If you want to use `sigpy_n_seq` function of `pypulseq`, here is an example code from the RF pulse test:

```python
    time_bw_product = 4
    slice_thickness = 3e-3  # Slice thickness
    flip_angle = np.pi / 2
    # Set system limits
    system = Opts(
        max_grad=32,
        grad_unit='mT/m',
        max_slew=130,
        slew_unit='T/m/s',
        rf_ringdown_time=30e-6,
        rf_dead_time=100e-6,
    )
    pulse_cfg = SigpyPulseOpts(
        pulse_type='slr',
        ptype='st',
        ftype='ls',
        d1=0.01,
        d2=0.01,
        cancel_alpha_phs=False,
        n_bands=3,
        band_sep=20,
        phs_0_pt='None',
    )
    rfp, gz, gzr = sigpy_n_seq(
        flip_angle=flip_angle,
        system=system,
        duration=3e-3,
        slice_thickness=slice_thickness,
        time_bw_product=4,
        return_gz=True,
        pulse_cfg=pulse_cfg,
        plot=True,
    )

    seq = pp.Sequence()
    seq.add_block(rfp)

```

There is no apodization input to this function. The documentation for this function is incorrect, I will update it soon.

If this function is not sufficient for your use case (e.g. you really need apodization), you can design your pulse externally as a waveform, and use `make_arbitrary_rf` function instead. Here is an SLR example:

```python
Trfmin       = 2e-3          # duration of one pulse segment [sec]
slice_thickness = 4e-3 # [m]
Nz = 20

n = ceil((Trfmin/dt)/4)*4
Trf = n*dt
tb = 8
bw = tb/Trf

signal = slr.dzrf(n=n, tb=tb, ptype='st', ftype='ls', d1=0.01, d2=0.01, cancel_alpha_phs=False)

# Create an RF event
rf_res, gz = make_arbitrary_rf(
    signal=signal, slice_thickness=slice_thickness*Nz,
    bandwidth=bw, flip_angle=pi/3, 
    system=system, return_gz=True, use="excitation"
    )

```

**Reply by qizixinge** [reply]
Thank you for your help. However, I still have two doubts during my coding. I did some tests on write_gre.py. 
First, I found “rf, gz, *_ = pp.make_sigpy_pulse.sigpy_n_seq(” success, but "rf, gz, gzr = pp.make_sigpy_pulse.sigpy_n_seq(" and "rf, gz, _ = pp.make_sigpy_pulse.sigpy_n_seq(" would result "ValueError: too many values to unpack". Perhaps the output of this function has more than three parameters as written in your file.
Second, I want to make "time_bw_product" to be 1.5 rather than 4 (my experiments or my pulse sequence design need it to be small as possible). However, I got the result as shown in the last figure. So how can I solve the problem?



![attempt1](https://github.com/user-attachments/assets/daaec5de-aae3-48a8-9c86-fd8599f395c3)

![result1](https://github.com/user-attachments/assets/f36cb64b-a11b-4e7b-813d-5052320c6e71)
![attempt2](https://github.com/user-attachments/assets/91fe6e90-af3e-4da3-81db-9f8511fbdc57)

![result2](https://github.com/user-attachments/assets/df888729-c8a0-4e0c-8cf1-d3ac8fed8a66)

**Reply by btasdelen** [reply]
You seem to be using an old version of PyPulseq. Please update to the latest version from PyPI and try again.

**Reply by qizixinge** [reply]
[writegreSLR.txt](https://github.com/user-attachments/files/18318842/writegreSLR.txt)
Thank you. But how can i update to the latest version from PyPI when running google colab? I tried "pip install git+https://github.com/imr-framework/pypulseq@dev", but still don't work. My original code on Google Colab is in the above txt file
![image](https://github.com/user-attachments/assets/a7a42e29-e7f1-4293-af33-1e8845ea7970)

**Reply by qizixinge** [reply]
And for the earliest version I used "pip install pypulseq"

**Reply by btasdelen** [reply, marked as answer]
Okay, I found the issue, but this is not an issue with `pypulseq`, but probably an issue with `sigpy`. Setting `time_bw_product=1.5` causes a negative value in a call to `firls` function, which, as stated in the error, only supports values between 0 and 1. In fact, you can only set the `time_bw_product` to 2 or higher. 

I don't know enough RF pulse design to say not having TBW less than 2 for SLR pulses is reasonable or not. I think you would have a better luck discussing this over [sigpy](https://github.com/mikgroup/sigpy). Overall, your code works if you set `time_bw_product=2`.
